### DISCUSSION TITLE
0050 XLS-50d: Aiming for a healthy distribution of (validator) infrastructure

### DISCUSSION THREAD

**Discussion by WietseWind** [XRPLF/XRPL-Standards#145] [XLS Proposals]
## The earlier version can be found by looking at the edit history.

**Comment by RichardAH** [comment]
Much needed. Consider adding an entry for who/which entity has physical custody of the server too. (Can be simply listed as "private individual" if no company.)

**Reply by WietseWind** [reply]
If one is OK with publicly disclosing this, a global, unique, accepted identifier would be great, in which case I propose a DUNS number.

**Reply by xVet** [reply]
What number would you give for private individuals ? or just DUNS = "private" and for companies a number ?

**Reply by WietseWind** [reply]
@xVet I'd say leave it out, "private" isn't a valid DUNS, obviously. Possibly another bool prop: "corporate". Is corporate, no DUNS = don't want to disclose. Non-corporate no DUNS = makes sense. Corporate + DUNS = ideal.

**Comment by wojake** [comment]
Since the network is based on overlapping trust, the addition of a section for "UNL management" (for validators) would improve this proposal.

**Comment by ihomp** [comment]
Bithomp now shows those values in the "Developer mode" (the checkbox on the top)

https://bithomp.com/validators 

<img width="996" alt="Screenshot 2023-11-16 at 14 39 10" src="https://github.com/XRPLF/XRPL-Standards/assets/19390499/7c4c1c59-fb9a-4345-aa65-941a7ce45d8d">
<img width="1067" alt="Screenshot 2023-11-16 at 14 39 24" src="https://github.com/XRPLF/XRPL-Standards/assets/19390499/32980f86-2c53-42c1-b235-858284863230">

**Reply by ihomp** [reply]
@xVet 24940 - looks like Hetzner cloud for me, no? =)

**Reply by xVet** [reply]
> @xVet 24940 - looks like Hetzner cloud for me, no? =) 

hmmm i'm not sure if i understood it correctly then, i thought if i do colocation in a datacenter then it's not cloud provided server? Thats why i put in "false". Isn't a cloud provided server any machine in a corporate DC that belongs to them and is managed by their software?

**Reply by WietseWind** [reply]
@xVet

Cloud = a machine you will never physically see/touch/know how to find imo. Good point btw. I updated the proposal to make the distinction between what is to be considered cloud and what not more clear.

Regarding provider, datacenter location, cloud opr not & ASN: the ASN is the infra (network, IP transit, traffic) provider. Dedicated, VPS, your own machine colocated: if (e.g. due to BGP routing table issues, DDoS, etc.) the provider has trouble (or their upstream peers do) all those machines can have connectivity issues at the same time.

Even your home/office ISP has an ASN.

Example: take the sample pic below, and say (on the right) AS3356 is the routing for an ISP/... and they have multiple datacenter locations (green dots) and there are dedi/colo/VPS machines there.

If AS3356 in this example isn't multi-homed & offline, or has a core router problem & redundancy fails, or ... - anyway: all machines *across their datacenters* are then in trouble. 

Hence my proposal to distribute also based on ASN.
<img width="1157" alt="image" src="https://github.com/XRPLF/XRPL-Standards/assets/4756161/ad4442d8-96b2-489f-817c-ba82953ead5f">

**Reply by xVet** [reply]
@WietseWind thank you, that makes it very clear now 🙏

**Reply by mvadari** [reply]
@ihomp how do you get the server addresses? It doesn't look like there's anything in the TOML to indicate that.

**Reply by ihomp** [reply]
@mvadari we use TOML properties described in this proposal ☝️  :
`network_asn` (integer) containing the ASN (autonomous system number) of the IP address the validator uses to connect out 
`server_location` (string) containing a written explanation of provider provided location details
`server_cloud` (boolean) if the server is running at a cloud provider

**Reply by mvadari** [reply]
@ihomp I was referring to the r-addresses, sorry.

**Reply by ihomp** [reply]
Ah, sorry, I thought it was something related to my screenshot :) 
The r-addresses we generate from public keys, we have shown that for years, the original idea and script we got from @RichardAH 
Basically, it's an address, that the validator operator can control with the validator keys. (The regular key is usually set, for easier access and control)
Haxau uses the same kind of validator's addresses in the governance game.
The function exists in our library:
`const address = BithompXRPL.Validator.classicAddressFromValidatorPK(publicKey);`
https://github.com/Bithomp/xrpl-api/blob/3b495a36f3d036dbe4972f3c6d1b442913d4bc31/src/validator.ts#L17

**Reply by WietseWind** [reply]
@mvadari https://gist.github.com/WietseWind/1f2b34ef6092e6f10762b5d526e89c86 as per @RichardAH's idea, from many years ago. Takes validator key, devices to keypair used for account, can then be rekeyed.

**Comment by lucbocahut** [comment]
We are currently evaluating cloud providers to host a validator.

re 1. Validator operator resource location selection 

What would be the best approach to build the current list of locations? Is there some public work on the main net topology? Do we need all validators to publish this information or can we build what we need from existing data? 

Thanks for your help

**Reply by WietseWind** [reply]
> Do we need all validators to publish this information

Preferably, at least: that's what I'm proposing, suggesting.

> build what we need from existing data

The data isn't there yet. I think it's safe to say that if you prevent using one of the bigger cloud/VPS providers (usual suspects: AWS, DigitalOcean, Hetzner, Google GPC, ...) you'll be fine.

**Comment by Tokeiito** [comment]
I think all information except asn is really useful for building automated topology map of ledger if majority of validators will adapt it. But I do not think that including asn is a good choice.
One of the main security considerations for validator, that it is anonymous in the network. Its IP address is not exposed and etc. Adding asn beats the purpose of [peer_private] and etc.

Revieling asn of gigants like gcp, azure maybe is not so much of the threat, but smaller hosting companies might have just a few of them and depending on configuration direct IP of validator might be leaked. At least in our case it is. Querying https://hackertarget.com/as-ip-lookup/ with our asn shows validator IP.

**Reply by WietseWind** [reply]
ASN brings one exactly zero inches closer to the actual node, most ASN's host a wide range of IP netblocks, and if peer_private is used, so should firewalls be, to limit incoming traffic. No specific IP can be targeted if the IP doesn't give away it's running rippled. 

Also: if there are corner cases where there's a micro ASN and giving away the IP blocks is deemed too much of a risk, then simply don't add the ASN to your stanza? Doesn't make considering publishing ASN a bad idea in general, just makes it a potential bad idea for the one corner case.

ASN is I'd say **one of the biggest infra centralisation risks**, while one of the **lowest privacy/attack risks**.

[a comment by an automated account is not shown]

**Reply by admin** [reply]
Already merged: https://github.com/XRPLF/XRPL-Standards/tree/master/XLS-0050-validator-toml-infra-details

[a comment by an automated account is not shown]
